runtime/compiler/control/CompilationThread.cpp

```diff
@@ -2641,35 +2644,89 @@ class ReleaseVMAccessAndAcquireMonitor
+ *            will attempt to release the monitor in its destructor.
+ */
+void TR::CompilationInfo::releaseCompMonitorUntilNotifiedOnCRMonitor(J9VMThread *vmThread) throw()
    {
-   J9JavaVM   *vm       = _jitConfig->javaVM;
-   J9VMThread *vmThread = vm->internalVMFunctions->currentVMThread(vm);
+   /* Acquire the CR monitor */
+   acquireCRMonitor();
 
-   if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestore))
-      TR_VerboseLog::writeLineLocked(TR_Vlog_CHECKPOINT_RESTORE, "Preparing for checkpoint");
+   /* Release the Comp Monitor before waiting */
+   releaseCompMonitor(vmThread);
+
+   /* Wait until notified */
+   waitOnCRMonitor();
+
+   /* Release CR Monitor and re-acquire the Comp Monitor */
+   releaseCRMonitor();
+   acquireCompMonitor(vmThread);
+   }
 
+bool TR::CompilationInfo::compileMethodsForCheckpoint(J9VMThread *vmThread)
    {
-   ReleaseVMAccessAndAcquireMonitor suspendCompThreadsForCheckpoint(vmThread, getCompilationMonitor());
+   if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestore))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_CHECKPOINT_RESTORE, "Preparing to compile methods for checkpoint");
 
-   if (TR::Options::_sleepMsBeforeCheckpoint)
+   /* Indicate to compilation threads that they should compile methods for checkpoint/restore */
+   setCompileMethodsForCheckpoint();
+
+   /* Queue compilation of interpreted methods */
+   try
       {
-      if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestore))
-         TR_VerboseLog::writeLineLocked(TR_Vlog_CHECKPOINT_RESTORE, "Sleeping for %d ms", TR::Options::_sleepMsBeforeCheckpoint);
+      TR_J9VMBase *fej9 = TR_J9VMBase::get(vmThread->javaVM->jitConfig, vmThread);
 
-      releaseCompMonitor(vmThread);
-      j9thread_sleep(static_cast<int64_t>(TR::Options::_sleepMsBeforeCheckpoint));
-      acquireCompMonitor(vmThread);
+      TR::RawAllocator rawAllocator(_jitConfig->javaVM);
+      J9::SegmentAllocator segmentAllocator(MEMORY_TYPE_JIT_SCRATCH_SPACE | MEMORY_TYPE_VIRTUAL, *_jitConfig->javaVM);
+      J9::SystemSegmentProvider regionSegmentProvider(
+         1 << 20,
+         1 << 20,
+         TR::Options::getScratchSpaceLimit(),
+         segmentAllocator,
+         rawAllocator
+         );
+      TR::Region compForCheckpointRegion(regionSegmentProvider, rawAllocator);
+
+      TR::CompileBeforeCheckpoint compileBeforeCheckpoint(compForCheckpointRegion, vmThread, fej9, this);
+      compileBeforeCheckpoint.collectAndCompileMethodsBeforeCheckpoint();
+      }
+   catch (const std::exception &e)
+      {
+      if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestore))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_CHECKPOINT_RESTORE, "Failed to collect methods for compilation before checkpoint");
       }
 
+   /* Determine whether to wait on the CR Monitor. */
+   while (getMethodQueueSize() && !shouldCheckpointBeInterrupted())
```

Comment:
It's ok to finish `compileMethodsForCheckpoint` because we then move on to `suspendCompThreadsForCheckpoint` [1] which only suspends the comp threads after the compilation is done.

[1] https://github.com/dsouzai/openj9/blob/6527d2435df80069110befcd936629cc35b6cd63/runtime/compiler/control/CompilationThread.cpp#L2723